Net/SMTP.php

```diff
@@ -1075,6 +1088,160 @@ public function authXOAuth2($uid, $token, $authz, $conn)
+
+    /**
+     * Authenticates the user using the SCRAM-SHA-256 method.
+     *
+     * @param string $uid   The userid to authenticate as.
+     * @param string $pwd   The password to authenticate with.
+     * @param string $authz The optional authorization proxy identifier.
+     *
+     * @return mixed Returns a PEAR_Error with an error message on any
+     *               kind of failure, or true on success.
+     * @since  1.11.0
+     */
+    protected function authScramSHA256($uid, $pwd, $authz = '')
+    {
+        $this->scram_sha_hash_algorithm = 'SCRAM-SHA-256';
+        return $this->authScramSHA($uid, $pwd, $authz);
+    }
+
+    /**
+     * Authenticates the user using the SCRAM-SHA-384 method.
+     *
+     * @param string $uid   The userid to authenticate as.
+     * @param string $pwd   The password to authenticate with.
+     * @param string $authz The optional authorization proxy identifier.
+     *
+     * @return mixed Returns a PEAR_Error with an error message on any
+     *               kind of failure, or true on success.
+     * @since  1.11.0
+     */
+    protected function authScramSHA384($uid, $pwd, $authz = '')
+    {
+        $this->scram_sha_hash_algorithm = 'SCRAM-SHA-384';
+        return $this->authScramSHA($uid, $pwd, $authz);
+    }
+
+    /**
+     * Authenticates the user using the SCRAM-SHA-512 method.
+     *
+     * @param string $uid   The userid to authenticate as.
+     * @param string $pwd   The password to authenticate with.
+     * @param string $authz The optional authorization proxy identifier.
+     *
+     * @return mixed Returns a PEAR_Error with an error message on any
+     *               kind of failure, or true on success.
+     * @since  1.11.0
+     */
+    protected function authScramSHA512($uid, $pwd, $authz = '')
+    {
+        $this->scram_sha_hash_algorithm = 'SCRAM-SHA-512';
+        return $this->authScramSHA($uid, $pwd, $authz);
+    }
+
+    /**
+     * Authenticates the user using the SCRAM-SHA method.
+     *
+     * @param string $uid   The userid to authenticate as.
+     * @param string $pwd   The password to authenticate with.
+     * @param string $authz The optional authorization proxy identifier.
+     *
+     * @return mixed Returns a PEAR_Error with an error message on any
+     *               kind of failure, or true on success.
+     * @since  1.11.0
+     */
+    protected function authScramSHA($uid, $pwd, $authz = '')
+    {
+        if (PEAR::isError($error = $this->put('AUTH', $this->scram_sha_hash_algorithm))) {
+            return $error;
+        }
+        /* 334: Continue authentication request */
+        if (PEAR::isError($error = $this->parseResponse(334))) {
+            /* 503: Error: already authenticated */
+            if ($this->code === 503) {
+                return true;
+            }
+            return $error;
+        }
+
+        $auth_sasl = new Auth_SASL;
+        $cram      = $auth_sasl->factory($this->scram_sha_hash_algorithm);
```

Comment:
Why not call `Auth_SASL::factory()` as a static function?

Reply:
Because, i only copy and paste the codeblock from `Net_SMTP->authCRAMMD5()`:
https://github.com/pear/Net_SMTP/blob/cfd963dc5cc73b4d64c7769e47dfa0f439dec536/Net/SMTP.php#L837

And on `Net_SMTP->authDigestMD5()` is the same:
https://github.com/pear/Net_SMTP/blob/cfd963dc5cc73b4d64c7769e47dfa0f439dec536/Net/SMTP.php#L784

So i don´t know whats best?
Should we leave the SCRAM one dynamic, or should we change all of the `Auth_SASL` based methods to static `Auth_SASL::factory()`?

Me opinion is always, never change a running system.
But you are right, this would be result in an error or a warning on PHP8, i think.
So when we change this, we have to change all of them.

What do you think?

Reply:
> Why not call `Auth_SASL::factory()` as a static function?
> 
> https://github.com/pear/Auth_SASL/blob/1f6f0eb4680027d76c555d550c55644fcc458c4f/Auth/SASL.php#L61

I changed all `Auth_SASL::factory` to static calls.